cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/runtimejob/DataprocRuntimeJobManager.java

```diff
@@ -0,0 +1,419 @@
+                                             .build());
+      RuntimeJobStatus runtimeJobStatus = getRuntimeJobStatus(job);
+
+      jobDetails = Optional.of(new RuntimeJobDetail(runtimeJobId, runtimeJobStatus));
+    } catch (ApiException e) {
+      // this may happen if job is manually deleted by user
+      if (e.getStatusCode().getCode() == StatusCode.Code.NOT_FOUND) {
+        LOG.warn("Dataproc job {} does not exist in project {}, region {}.", jobId, projectId, region);
+      } else {
+        throw new Exception(String.format("Error while getting details for job %s on cluster %s.",
+                                          jobId, clusterName), e);
+      }
+    }
+    return jobDetails;
+  }
+
+  @Override
+  public List<RuntimeJobDetail> list() throws Exception {
+    Set<String> filters = new HashSet<>();
+    // Dataproc jobs can be filtered by status.state filter. In this case we only want ACTIVE jobs.
+    filters.add("status.state=ACTIVE");
+    // Filter by labels that were added to the job when this runtime job manager submitted dataproc job. Note that
+    // dataproc only supports AND filter.
+    for (Map.Entry<String, String> entry : labels.entrySet()) {
+      filters.add("labels." + entry.getKey() + "=" + entry.getValue());
+    }
+    String jobFilter = Joiner.on(" AND ").join(filters);
+
+    LOG.info("Getting a list of jobs under project {}, region {}, cluster {} with filter {}.", projectId, region,
+             clusterName, jobFilter);
+    JobControllerClient.ListJobsPagedResponse listJobsPagedResponse =
+      jobControllerClient.listJobs(ListJobsRequest.newBuilder()
+                                     .setProjectId(projectId).setRegion(region).setClusterName(clusterName)
+                                     .setFilter(jobFilter).build());
+
+    Iterator<Job> jobsItor = listJobsPagedResponse.iterateAll().iterator();
+    List<RuntimeJobDetail> jobsDetail = new ArrayList<>();
+    while (jobsItor.hasNext()) {
+      Job job = jobsItor.next();
+      jobsDetail.add(new RuntimeJobDetail(new RuntimeJobId(job.getReference().getJobId()), getRuntimeJobStatus(job)));
+    }
+    return jobsDetail;
+  }
+
+  @Override
+  public void stop(RuntimeJobId runtimeJobId) throws Exception {
+    Optional<RuntimeJobDetail> jobDetail = getDetail(runtimeJobId);
+    // if the job does not exist, it can be safely assume that job has been deleted. Hence has reached terminal state.
+    if (!jobDetail.isPresent()) {
+      return;
+    }
+    // stop dataproc job
+    Job stoppedJob = stopJob(runtimeJobId);
+
+    // delete gcs path for the job
+    if (stoppedJob != null) {
+      String runRootPath = getPath(CDAP_GCS_ROOT, stoppedJob.getHadoopJob().getPropertiesMap().get(CDAP_RUNTIME_RUNID));
+      deleteGCSPath(runRootPath);
+    }
+  }
+
+  @Override
+  public void kill(RuntimeJobId runtimeJobId) throws Exception {
+    stop(runtimeJobId);
+  }
+
+  @Override
+  public void destroy() {
+    jobControllerClient.close();
```

Comment:
Do you need to close the `storageClient` as well?

Reply:
no storage client does not have close method.